From c0e2bab1bbc21917c3afb5f1c76e65e1e388b9b3 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Fri, 24 Feb 2017 09:26:57 +0800
Subject: [PATCH 1642/5242] MLK-14257 video: mxsfb: fix a deadlock issue for
 overlay fb

commit  e559d85054f492ce701a8d26c84963bec02e13af from
https://source.codeaurora.org/external/imx/linux-imx.git

The overlayfb_enable() function would call fb0 info lock,
so in some overlay functions which may call overlay_enable()
should not lock the fb0 info.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit ec53f89c7f702aa56e7fe6f360e23582ed78302f)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxsfb.c |   10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/drivers/video/fbdev/mxsfb.c b/drivers/video/fbdev/mxsfb.c
index a9baa4e..57722ad 100644
--- a/drivers/video/fbdev/mxsfb.c
+++ b/drivers/video/fbdev/mxsfb.c
@@ -1859,6 +1859,7 @@ static int overlayfb_set_par(struct fb_info *info)
 		clk_enable_axi(fbi);
 		clk_enable_disp_axi(fbi);
 	}
+	unlock_fb_info(fbi->fb_info);
 
 	if (ofb->blank_state == FB_BLANK_UNBLANK)
 		ofb->ops->disable(ofb);
@@ -1868,12 +1869,14 @@ static int overlayfb_set_par(struct fb_info *info)
 	if (ofb->blank_state == FB_BLANK_UNBLANK)
 		ofb->ops->enable(ofb);
 
+	if (!lock_fb_info(fbi->fb_info))
+		return -EINVAL;
+
 	if (fbi->cur_blank != FB_BLANK_UNBLANK) {
 		clk_disable_disp_axi(fbi);
 		clk_disable_axi(fbi);
 		clk_disable_pix(fbi);
 	}
-
 	unlock_fb_info(fbi->fb_info);
 
 	if ((var->activate & FB_ACTIVATE_FORCE) &&
@@ -1899,6 +1902,7 @@ static int overlayfb_blank(int blank, struct fb_info *info)
 		clk_enable_axi(fbi);
 		clk_enable_disp_axi(fbi);
 	}
+	unlock_fb_info(fbi->fb_info);
 
 	switch (blank) {
 	case FB_BLANK_POWERDOWN:
@@ -1912,12 +1916,14 @@ static int overlayfb_blank(int blank, struct fb_info *info)
 		break;
 	}
 
+	if (!lock_fb_info(fbi->fb_info))
+		return -EINVAL;
+
 	if (fbi->cur_blank != FB_BLANK_UNBLANK) {
 		clk_disable_disp_axi(fbi);
 		clk_disable_axi(fbi);
 		clk_disable_pix(fbi);
 	}
-
 	unlock_fb_info(fbi->fb_info);
 
 	ofb->blank_state = blank;
-- 
1.7.9.5

